webapi/src/test/scala/org/knora/webapi/messages/util/search/gravsearch/prequery/NonTriplestoreSpecificGravsearchToPrequeryTransformerSpec.scala

```diff
@@ -1319,8 +1320,214 @@ class NonTriplestoreSpecificGravsearchToPrequeryTransformerSpec extends CoreSpec
         useDistinct = true
     )
 
+    val queryWithOptional: String =
+        """
+          |PREFIX beol: <http://0.0.0.0:3333/ontology/0801/beol/simple/v2#>
+          |PREFIX knora-api: <http://api.knora.org/ontology/knora-api/simple/v2#>
+          |PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
+          |
+          |CONSTRUCT {
+          |    ?document knora-api:isMainResource true .
+          |} WHERE {
+          |    ?document rdf:type beol:writtenSource .
+          |
+          |    OPTIONAL {
+          |    ?document beol:hasRecipient ?recipient .
+          |
+          |    ?recipient beol:hasFamilyName ?familyName .
+          |
+          |    FILTER knora-api:matchText(?familyName, "Bernoulli")
+          |}
+          |}
+                """.stripMargin
+
+    val TransformedQueryWithOptional: SelectQuery = SelectQuery(
+        variables = List(QueryVariable(variableName = "document")),
+        offset = 0,
+        groupBy = Vector(QueryVariable(variableName = "document")),
+        orderBy = Vector(OrderCriterion(
+            queryVariable = QueryVariable(variableName = "document"),
+            isAscending = true
+        )),
+        whereClause = WhereClause(
+            patterns = ArrayBuffer(
+                StatementPattern(
+                    subj = QueryVariable(variableName = "document"),
+                    pred = IriRef(
+                        iri = "http://www.knora.org/ontology/knora-base#isDeleted".toSmartIri,
+                        propertyPathOperator = None
+                    ),
+                    obj = XsdLiteral(
+                        value = "false",
+                        datatype = "http://www.w3.org/2001/XMLSchema#boolean".toSmartIri
+                    ),
+                    namedGraph = Some(IriRef(
+                        iri = "http://www.knora.org/explicit".toSmartIri,
+                        propertyPathOperator = None
+                    ))
+                ),
+                // This statement must not be removed by AbstractPrequeryGenerator.removeEntitiesInferredFromProperty
+                // because the property from which its type can be inferred is in an optional. Without this statement,
+                // the type beol:basicLetter (inferred from property beol:hasRecipient) would be considered for ?document.
+                StatementPattern(
```

Comment:
@benjamingeer If we remove the rdf:type statement with object beol:writtenSource. I would have also expected it to consider any kind of resource, but Triplestore infers the type to be beol:basicLetter from the OPTIONAL statement.

Reply:
> Triplestore infers the type to be beol:basicLetter from the OPTIONAL statement.

Do you mean that all the results are instances of `beol:basicLetter`? How many results does it return? Is there more than one page?

Reply:
Ok, let me add an end to end test for this.

Reply:
I just ran this query in the Fuseki dashboard (http://localhost:3030/dataset.html?tab=query&ds=/knora-test):

```sparql
PREFIX beol: <http://www.knora.org/ontology/0801/beol#>
PREFIX knora-base: <http://www.knora.org/ontology/knora-base#>

SELECT ?document WHERE {
  ?document knora-base:isDeleted false .

  OPTIONAL {
    ?document beol:hasRecipient ?recipient .
    ?recipient beol:hasFamilyName ?familyName .
  }
}
```

It returns 30,435 results, including resources of different types (e.g. `incunabula:book`) as well as values.

Reply:
When I run this, I get 371,400 results including values!
But if you run it without '?document knora-base:isDeleted false .' you get only one result. I mean following query returns one result.
```
PREFIX beol: <http://www.knora.org/ontology/0801/beol#>
PREFIX knora-base: <http://www.knora.org/ontology/knora-base#>

SELECT ?document WHERE {
  OPTIONAL {
    ?document beol:hasRecipient ?recipient .
    ?recipient beol:hasFamilyName ?familyName .
  }
}
```

Reply:
When we remove the rdf:type statements for entities that can be inferred from properties, the isDeleted statements are not generated either.

Reply:
OK, I think I understand it now.

This query returns nothing:

```sparql
SELECT ?document WHERE { }
```

That's because the WHERE clause doesn't bind anything to `?document`. Now we add an `OPTIONAL`:

```sparql
SELECT ?document WHERE {
  OPTIONAL {
    ?document beol:hasRecipient ?recipient .
  }
}
```

The only thing an `OPTIONAL` can do is **add** data to the results. So the result of this query should be the same as the first one, plus whatever the `OPTIONAL` found. Does that make sense?

Reply:
> The only thing an `OPTIONAL` can do is **add** data to the results. So the result of this query should be the same as the first one, plus whatever the `OPTIONAL` found. Does that make sense?

yes, it does.

Reply:
> The only thing an `OPTIONAL` can do is **add** data to the results. So the result of this query should be the same as the first one, plus whatever the `OPTIONAL` found. Does that make sense?

yes, it does. So if we want the query to return something meaningful, we should not remove the rdfs:type statement in this case.